geo/src/algorithm/concave_hull.rs

```diff
@@ -0,0 +1,432 @@
+    /// This implementation is inspired by https://github.com/mapbox/concaveman
+    /// and also uses ideas from the following paper:
+    /// www.iis.sinica.edu.tw/page/jise/2012/201205_10.pdf
+    ///
+    /// # Example
+    /// ```
+    /// use geo::{line_string, polygon};
+    /// use geo::algorithm::concave_hull::ConcaveHull;
+    ///
+    /// // a square shape
+    /// let poly = polygon![
+    ///     (x: 0.0, y: 0.0),
+    ///     (x: 4.0, y: 0.0),
+    ///     (x: 4.0, y: 4.0),
+    ///     (x: 0.0, y: 4.0),
+    /// ];
+    ///
+    /// // The correct concave hull coordinates
+    /// let correct_hull = line_string![
+    ///     (x: 4.0, y: 0.0),
+    ///     (x: 4.0, y: 4.0),
+    ///     (x: 0.0, y: 4.0),
+    ///     (x: 0.0, y: 0.0),
+    ///     (x: 4.0, y: 0.0),
+    /// ];
+    ///
+    /// let res = poly.concave_hull(2.0);
+    /// assert_eq!(res.exterior(), &correct_hull);
+    /// ```
+    fn concave_hull(&self, concavity: T) -> Polygon<T>
+    where
+        T: Float + RTreeNum;
+}
+
+impl<T> ConcaveHull<T> for Polygon<T>
+where
+    T: Float + RTreeNum,
+{
+    fn concave_hull(&self, concavity: T) -> Polygon<T> {
+        let points: Vec<_> = self.exterior().points_iter().collect();
+        Polygon::new(LineString::from(concave_hull(&points, concavity)), vec![])
+    }
+}
+
+impl<T> ConcaveHull<T> for MultiPolygon<T>
+where
+    T: Float + RTreeNum,
+{
+    fn concave_hull(&self, concavity: T) -> Polygon<T> {
+        let aggregated: Vec<Point<T>> = self
+            .0
+            .iter()
+            .flat_map(|elem| elem.exterior().0.iter().map(|c| Point(*c)))
+            .collect();
+        Polygon::new(
+            LineString::from(concave_hull(&aggregated, concavity)),
+            vec![],
+        )
+    }
+}
+
+impl<T> ConcaveHull<T> for LineString<T>
+where
+    T: Float + RTreeNum,
+{
+    fn concave_hull(&self, concavity: T) -> Polygon<T> {
+        Polygon::new(
+            LineString::from(concave_hull(&self.clone().into_points(), concavity)),
+            vec![],
+        )
+    }
+}
+
+impl<T> ConcaveHull<T> for MultiLineString<T>
+where
+    T: Float + RTreeNum,
+{
+    fn concave_hull(&self, concavity: T) -> Polygon<T> {
+        let aggregated: Vec<Point<T>> = self
+            .0
+            .iter()
+            .flat_map(|elem| elem.clone().into_points())
+            .collect();
+        Polygon::new(
+            LineString::from(concave_hull(&aggregated, concavity)),
+            vec![],
+        )
+    }
+}
+
+impl<T> ConcaveHull<T> for MultiPoint<T>
+where
+    T: Float + RTreeNum,
+{
+    fn concave_hull(&self, concavity: T) -> Polygon<T> {
+        Polygon::new(
+            LineString::from(concave_hull(&self.0.clone(), concavity)),
+            vec![],
+        )
+    }
+}
+
+fn find_point_closest_to_line<T>(
+    interior_points_tree: &RTree<Point<T>>,
+    line: Line<T>,
+    max_dist: T,
+    edge_length: T,
+    concavity: T,
+    line_tree: &RTree<Line<T>>
+) -> Option<Point<T>>
```

Comment:
Another minor suggestion:  all the `Point<T>` could be `Coordinate<T>`.  The idea is that `Point` may in the future be more than just the coordinates, and the algos that directly work on coordinates shouldn't need to use it.  This should also simplify some of the impls, as the geometry types store `Coordinate`s

Reply:
Since the current `RTreeObject` implementation's don't include `Coordinate` would it be sensible to keep it as `Point`  because otherwise you get casts everywhere as you convert from `Coordinate` to `Point` when you need RTrees and you convert back when you don't.

Reply:
If you're feeling bold you could implement the trait for Coordinate.

I think like this in `coordinate.rs`
```
 74 #[cfg(feature = "rstar")]
 75 impl<T> ::rstar::Point for Coordinate<T>
 76 where 
 77     T: ::num_traits::Float + ::rstar::RTreeNum,
 78 {
 79     type Scalar = T;
 80 
 81     const DIMENSIONS: usize = 2;
 82 
 83     fn generate(generator: impl Fn(usize) -> Self::Scalar) -> Self {
 84         Coordinate { x: generator(0), y: generator(1) }
 85     }         
 86 
 87     fn nth(&self, index: usize) -> Self::Scalar {
 88         match index {
 89             0 => self.x, 
 90             1 => self.y,
 91             _ => unreachable!(),
 92         } 
 93     } 
 94 
 95     fn nth_mut(&mut self, index: usize) -> &mut Self::Scalar {
 96         match index {
 97             0 => &mut self.x,
 98             1 => &mut self.y,
 99             _ => unreachable!(),
100         } 
101     }     
102 }
```

I think that this is an implementation detail that doesn't affect the user facing API, so I don't feel strongly about it, but I do agree with @rmanoka's sentiment that we should be preferring Coord whenever possible. 

Unfortunately I notice that the Rstar impl's on linestring use Points (not Coord), but I'm not sure if that affects you here.

Reply:
Ooh, I'll give this a try!